docs/foreign-keys.md

```diff
@@ -0,0 +1,199 @@
+    "allOf": [
+        { "$ref": "#/definitions/abstractBaseField" },
+        { "$ref": "#/definitions/abstractRequirableField" }
+    ],
+    "title": "Record-to-Record Relationship",
+    "properties": {
+        "fieldTitle": {
+            "title": "Relationship Title",
+            "type": "string",
+            "minLength": 1
+        },
+        "referenceTarget": {
+            "title": "RecordType of the related Record",
+            "type": "string",
+            "format": "select",
+            "enumSource": [["Will be dynamically overwritten"]]
+        },
+        "fieldType": {
+            "options": {
+                "hidden": true
+            },
+            "type": "string",
+            "enum": ["reference"]
+        }
+    }
+}
+```
+
+Then when loading a form for the schema editor, the `schemas-service` script
+could populate the `referenceTarget` with all available RecordType UUIDs.
+Additional care would have to be taken to make sure that the data collector
+view would only show options for Records that belong to this RecordType, but
+that filtering should be trivial. 
+
+#### Requirements
+
+For Record-to-Record foreign keys to be workable, they need to permit three basic
+operations:
+
+1. **Lookups**: Foreign key relationships need to support nested search and
+   filtering to match the rest of the Grout field types.
+2. **Validation**: When an incoming record includes a foreign key, the validation function
+   (`grout.models.SchemaModel.validate_schema`) needs to be able to check
+   whether the referent exists in the database.
+3. **Indexing**: Database lookups based on foreign keys must be performant.
+   Without a way to index foreign key relationships, lookups may not be
+   reasonable in large production systems. 
+
+#### A brief note on null pointers
+
+One operation that didn't make this list is **Resolution of null pointers**.
+Null pointers are a classic problem with non-relational data stores: if document
+A points to document B, and then document B is removed from the system, NoSQL
+databases offer no system for automatically resolving the null reference
+(whether by raising an error, setting a default, or simply storing a `NULL`
+value for the foreign key). In the example given above, we could imagine a user
+deleting an Event referenced by a Poster, and then the UUID stored in
+Poster.event would not reference a coherent entity in the database.
+
+Null pointers are indeed a problem with Record-to-Record foreign keys, but this
+problem is already endemic to Grout: if a field is removed from a schema during
+an upgrade to a new version (i.e. a definition of a new RecordSchema), Grout
+silently permits the field to remain but become unaccessible. From the reverse
+perspective, if a field is _added_ in the new schema, old records will not store
+values for that field, and so will not appear in the relevant filters.
+
+While not precisely a "null pointer," the issue is functionally the same: some data
```

Comment:
That's a good point! When I was first investigating the data model, I put some thought into versioning for Records (notes here: https://github.com/azavea/ashlar/issues/80). Reading back on that, it seems to imply that Records and RecordTypes cannot be deleted, they can simply have an `active` value of `True` or `False`. I wonder if this is already exposed in the data collector frontend? If it's truly impossible for a user to fully delete a Record (without calling a manual SQL update), would that address some of your concern?

Reply:
Yeah, I think not allowing deletions is potentially a good way to address this. If deletion is not allowed, a Record would always have the data it was created with available, even if that data was marked deleted. Versioning would be even better, because then if a referent gets changed, any records referring to it would still have access to the original data. As an alternative to `ON DELETE` I kind of prefer this because I think it aligns better with the data model for Schemas, and because it keeps deletions constant-time. With `ON DELETE`, a deletion could potentially be very costly if the Record being deleted was referred to by many other Records.